src/common/types/column_data_allocator.cpp

```diff
@@ -154,6 +159,44 @@ data_ptr_t ColumnDataAllocator::GetDataPointer(ChunkManagementState &state, uint
 	return state.handles[block_id].Ptr() + offset;
 }
 
+void ColumnDataAllocator::UnswizzlePointers(ChunkManagementState &state, Vector &result, uint16_t v_offset,
+                                            uint16_t count, uint32_t block_id, uint32_t offset) {
+	D_ASSERT(result.GetType().InternalType() == PhysicalType::VARCHAR);
+	lock_guard<mutex> guard(lock);
+
+	auto &validity = FlatVector::Validity(result);
+	auto strings = FlatVector::GetData<string_t>(result);
+
+	// find first non-inlined string
+	auto i = v_offset;
+	const auto end = v_offset + count;
+	for (; i < end; i++) {
+		if (!validity.RowIsValid(i)) {
+			continue;
+		}
+		if (!strings[i].IsInlined()) {
+			break;
+		}
+	}
+	// at least one string must be non-inlined, otherwise this function should not be called
+	D_ASSERT(i < end);
+
+	auto base_ptr = (char *)GetDataPointer(state, block_id, offset);
+	if (strings[i].GetDataUnsafe() == base_ptr) {
+		// pointers are still valid
+		return;
+	}
+
+	// pointer mismatch! pointers are invalid, set them correctly
+	for (; i < end; i++) {
+		if (!validity.RowIsValid(i)) {
+			continue;
+		}
+		strings[i].SetPointer(base_ptr);
```

Comment:
Should this not happen only if the string is not inlined? What if we have inlined and non-inlined strings mixed?

Reply:
Good catch. That is also the reason for the few tests that are failing.

However, by running the tests with both of the flags as you suggested, I found a bug in `test/sql/copy/parquet/writer/test_parquet_write_complex.test`. It seems like a pin is not being held for long enough to ensure that the `string_t` pointers in the hashmap for the dictionary compression are still valid.

I will try to see if I can fix it.

Reply:
I fixed this by adding a `dictionary_heap` StringHeap object in the parquet writer, which uses the `BufferAllocator`.
Strings inserted into the `dictionary` add their strings to this heap. This could possibly be done in a zero-copy way, but that would require refactoring how the parquet writer iterates over the chunk in a `ChunkCollection`, which should be done in a separate PR.

Currently, it uses this pattern: `for (auto &chunk : buffer.Chunks())`, but it could use the `Scan` functions instead and keep all the blocks pinned.

Reply:
This also got me thinking, should we combine our `DESTROY_UNPINNED_BLOCKS` and `DISABLE_STRING_INLINE` CI runs? Together, these flags find more bugs than alone.